Small tidying work for coherently invertible maps (#1024)
UniMath/agda-unimath: src/foundation-core/commuting-squares-of-homotopies.lagda.md

      ( right ·r F)
      ( bottom ·r F)
  right-whisker-comp-coherence-square-homotopies top left right bottom F α =
    α ·r F
```

### Concatenating homotopies of edges and coherences of commuting squares of homotopies

Consider a commuting square of homotopies and a homotopy of one of the four
sides with another homotopy, as for example in the diagram below:

```text
             top
       a ---------> b
       |           | |
  left |     right |~| right'
       ∨           ∨ ∨
       c ---------> d.
           bottom
```

Then any homotopy witnessing that the square commutes can be concatenated with
the homotopy on the side, to obtain a new commuting square of homotopies.

**Note.** To avoid cyclic module dependencies we will give direct proofs that
concatenating homotopies of edges of a square with the coherence of its
commutativity is an equivalence.

#### Concatenating homotopies of the top edge with a coherence of a commuting square of homotopies

Consider a commuting diagram of homotopies

```text
           top'
         ------->
       f -------> g
       |   top    |
  left |          | right
       ∨          ∨
       h -------> i.
          bottom
```

> with a homotopy `top ~ top'`. Then we get an equivalence
> 
> ```text
>            top                             top'
>        f -------> g                    f -------> g
>        |          |                    |          |
>   left |          | right    ≃    left |          | right
>        ∨          ∨                    ∨          ∨
>        h -------> i                    h -------> i.
>           bottom                          bottom
> ```

Reviewer: ```suggestion
with a homotopy `top ~ top'`. Then we get map

```text
           top                             top'
       f -------> g                    f -------> g
       |          |                    |          |
  left |          | right    →    left |          | right
       ∨          ∨                    ∨          ∨
       h -------> i                    h -------> i.
          bottom                          bottom
```
```

```agda
module _
  {l1 l2 : Level} {A : UU l1} {B : A → UU l2} {f g h i : (x : A) → B x}
  (top : f ~ g) (left : f ~ h) (right : g ~ i) (bottom : h ~ i)
  {top' : f ~ g} (s : top ~ top')
  where

  concat-top-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies top' left right bottom
  concat-top-homotopy-coherence-square-homotopies H x =
    concat-top-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)

  inv-concat-top-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top' left right bottom →
    coherence-square-homotopies top left right bottom
  inv-concat-top-homotopy-coherence-square-homotopies H x =
    inv-concat-top-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)
```

We record that this construction is an equivalence in

> [`foundation.commuting-squares-of-homotopies`](foundation.commuting-squares-of-homotopies.md).

Reviewer: Probably state this sentence before the code block

#### Concatenating homotopies of the left edge with a coherence of a commuting square of homotopies

Consider a commuting diagram of homotopies

```text
              top
         f -------> g
        | |         |
  left' | | left    | right
        ∨ ∨         ∨
         h -------> i.
            bottom
```

> with a homotopy `left ~ left'`. Then we get an equivalence
> 
> ```text
>            top                              top
>        f -------> g                     f -------> g
>        |          |                     |          |
>   left |          | right    ≃    left' |          | right
>        ∨          ∨                     ∨          ∨
>        h -------> i                     h -------> i.
>           bottom                           bottom
> ```

Reviewer: ```suggestion
with a homotopy `left ~ left'`. Then we get maps back and forth

```text
           top                              top
       f -------> g                     f -------> g
       |          |                     |          |
  left |          | right    ↔    left' |          | right
       ∨          ∨                     ∨          ∨
       h -------> i                     h -------> i.
          bottom                           bottom
```
```

    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies top left' right bottom
  concat-left-homotopy-coherence-square-homotopies H x =
    concat-left-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)

  inv-concat-left-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top left' right bottom →
    coherence-square-homotopies top left right bottom
  inv-concat-left-homotopy-coherence-square-homotopies H x =
    inv-concat-left-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)
```

We record that this construction is an equivalence in
[`foundation.commuting-squares-of-homotopies`](foundation.commuting-squares-of-homotopies.md).

#### Concatenating homotopies of the right edge with a coherence of a commuting square of homotopies

Consider a commuting diagram of homotopies

```text
            top
       f -------> g
       |         | |
  left |   right | | right'
       ∨         ∨ ∨
       h -------> i.
          bottom
```

> with a homotopy `right ~ right'`. Then we get an equivalence
> 
> ```text
>            top                             top
>        f -------> g                    f -------> g
>        |          |                    |          |
>   left |          | right    ≃    left |          | right'
>        ∨          ∨                    ∨          ∨
>        h -------> i                    h -------> i.
>           bottom                          bottom
> ```

Reviewer: ```suggestion
with a homotopy `right ~ right'`. Then we get maps back and forth

```text
           top                             top
       f -------> g                    f -------> g
       |          |                    |          |
  left |          | right    ↔    left |          | right'
       ∨          ∨                    ∨          ∨
       h -------> i                    h -------> i.
          bottom                          bottom
```
```

    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies top left right' bottom
  concat-right-homotopy-coherence-square-homotopies H x =
    concat-right-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)

  inv-concat-right-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top left right' bottom →
    coherence-square-homotopies top left right bottom
  inv-concat-right-homotopy-coherence-square-homotopies H x =
    inv-concat-right-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)
```

We record that this construction is an equivalence in
[`foundation.commuting-squares-of-homotopies`](foundation.commuting-squares-of-homotopies.md).

#### Concatenating homotopies of the bottom edge with a coherence of a commuting square of homotopies

Consider a commuting diagram of homotopies

```text
            top
       f -------> g
       |          |
  left |          | right
       ∨  bottom  ∨
       h -------> i.
         ------->
          bottom'
```

> with a homotopy `bottom ~ bottom'`. Then we get an equivalence
> 
> ```text
>            top                             top
>        f -------> g                    f -------> g
>        |          |                    |          |
>   left |          | right    ≃    left |          | right
>        ∨          ∨                    ∨          ∨
>        h -------> i                    h -------> i.
>           bottom                          bottom'
> ```

Reviewer: ```suggestion
with a homotopy `bottom ~ bottom'`. Then we get maps back and forth

```text
           top                             top
       f -------> g                    f -------> g
       |          |                    |          |
  left |          | right    ↔    left |          | right
       ∨          ∨                    ∨          ∨
       h -------> i                    h -------> i.
          bottom                          bottom'
```
```

```agda
module _
  {l1 l2 : Level} {A : UU l1} {B : A → UU l2} {f g h i : (x : A) → B x}
  (top : f ~ g) (left : f ~ h) (right : g ~ i) (bottom : h ~ i)
  {bottom' : h ~ i} (s : bottom ~ bottom')
  where

  concat-bottom-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies top left right bottom'
  concat-bottom-homotopy-coherence-square-homotopies H x =
    concat-bottom-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)

  inv-concat-bottom-homotopy-coherence-square-homotopies :
    coherence-square-homotopies top left right bottom' →
    coherence-square-homotopies top left right bottom
  inv-concat-bottom-homotopy-coherence-square-homotopies H x =
    inv-concat-bottom-identification-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( s x)
      ( H x)
```

We record that this construction is an equivalence in
[`foundation.commuting-squares-of-homotopies`](foundation.commuting-squares-of-homotopies.md).

> ### Whiskering and splicing coherences of commuting squares of homotopies

Reviewer: ```suggestion
### Whiskering and splicing coherences of commuting squares of homotopies with respect to concatenation of homotopies
```

   commuting square of homotopies

   ```text
                      top
              f --------------> g
              |                 |
    left ∙h H |                 | right
              ∨                 ∨
              u --------------> i.
                 H⁻¹ ∙h bottom
   ```

   More precisely, we have an equivalence

   ```text
     (left ∙h bottom ~ top ∙h right) ≃ ((left ∙h H) ∙h (H⁻¹ ∙h bottom) ~ top ∙h right).
   ```

   Similarly, we have an equivalence

   ```text
     (left ∙h bottom ~ top ∙h right) ≃ ((left ∙h H⁻¹) ∙h (H ∙h bottom) ~ top ∙h right).
   ```

4. Splicing a homotopy `H : g ~ u` and its inverse into the middle gives a
   commuting square of homotopies

   ```text
            top ∙h H
          f --------> u
          |           |
     left |           | H⁻¹ ∙h right
          ∨           ∨
          h --------> i.
             bottom
   ```

   More precisely, we have an equivalence

   ```text
     (left ∙h bottom ~ top ∙h right) ≃ (left ∙h bottom ~ (top ∙h H) ∙h (H⁻¹ ∙h right)).
   ```

   Similarly, we have an equivalence

   ```text
     (left ∙h bottom ~ top ∙h right) ≃ (left ∙h bottom ~ (top ∙h H⁻¹) ∙h (H ∙h right)).
   ```

> These operations are useful in proofs involving path algebra, because taking
> `equiv-right-whisker-concat-coherence-square-identicications` as an example, it
> provides us with two maps: the forward direction states
> `(H ∙h r ~ K ∙h s) → (H ∙h (r ∙h t)) ~ K ∙h (s ∙h t))`, which allows one to
> append an homotopy without needing to reassociate on the right, and the
> backwards direction conversely allows one to cancel out a homotopy in
> parentheses.

Reviewer: ```suggestion
These operations are useful in proofs involving homotopy algebra, because taking
`right-whisker-concat-coherence-square-homotopies` as an example, it
provides us with two maps: the forward direction states
`(H ∙h r ~ K ∙h s) → (H ∙h (r ∙h t)) ~ K ∙h (s ∙h t))`, which allows one to
append a homotopy without needing to reassociate on the right, and the
backwards direction conversely allows one to cancel out a homotopy in
parentheses.
```

This text has been copied over from the analogous file about squares of identifications, but not all adjustments to the setting of homotopies are made.

#### Left whiskering coherences of commuting squares of homotopies

> For any homotopy `H : u ~ f` we obtain an equivalence
> 
> ```text
>            top                                H ∙h top
>        f -------> g                         u -------> g
>        |          |                         |          |
>   left |          | right    ≃    H ∙h left |          | right
>        ∨          ∨                         ∨          ∨
>        h -------> i                         h -------> i
>           bottom                               bottom
> ```
> 
> of coherences of commuting squares of homotopies.

Reviewer: ```suggestion
For any homotopy `H : u ~ f` we obtain maps back and forth

```text
           top                                H ∙h top
       f -------> g                         u -------> g
       |          |                         |          |
  left |          | right    ↔    H ∙h left |          | right
       ∨          ∨                         ∨          ∨
       h -------> i                         h -------> i
          bottom                               bottom
```

We show in ... that these maps are equivalences.

of coherences of commuting squares of homotopies.
```

  {l1 l2 : Level} {A : UU l1} {B : A → UU l2} {f g h i : (x : A) → B x}
  where

  left-whisker-concat-coherence-square-homotopies :
    {u : (x : A) → B x} (H : u ~ f)
    (top : f ~ g) (left : f ~ h) (right : g ~ i) (bottom : h ~ i) →
    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies (H ∙h top) (H ∙h left) right bottom
  left-whisker-concat-coherence-square-homotopies
    H top left right bottom coh x =
    left-whisker-concat-coherence-square-identifications
      ( H x)
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( coh x)

  left-unwhisker-concat-coherence-square-homotopies :
    {u : (x : A) → B x} (H : u ~ f)
    (top : f ~ g) (left : f ~ h) (right : g ~ i) (bottom : h ~ i) →
    coherence-square-homotopies (H ∙h top) (H ∙h left) right bottom →
    coherence-square-homotopies top left right bottom
  left-unwhisker-concat-coherence-square-homotopies
    H top left right bottom coh x =
    left-unwhisker-concat-coherence-square-identifications
      ( H x)
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( coh x)
```

#### Right whiskering coherences of commuting squares of homotopies

> For any homotopy `H : i ~ u` we obtain an equivalence
> 
> ```text
>            top                                 top
>        f -------> g                     f ------------> g
>        |          |                     |               |
>   left |          | right    ≃     left |               | right ∙h H
>        ∨          ∨                     ∨               ∨
>        h -------> i                     h ------------> i
>           bottom                           bottom ∙h H
> ```
> 
> of coherences of commuting squares of homotopies.

Reviewer: ````suggestion
For any homotopy `H : i ~ u` we obtain maps back and forth

```text
           top                                 top
       f -------> g                     f ------------> g
       |          |                     |               |
  left |          | right    ↔     left |               | right ∙h H
       ∨          ∨                     ∨               ∨
       h -------> i                     h ------------> i
          bottom                           bottom ∙h H
```

of coherences of commuting squares of homotopies.
````

We show in ... that these maps are equivalences.

    coherence-square-homotopies top left (right ∙h H) (bottom ∙h H) →
    coherence-square-homotopies top left right bottom
  right-unwhisker-cohernece-square-homotopies H coh x =
    right-unwhisker-cohernece-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( H x)
      ( coh x)
```

### Double whiskering of commuting squares of homotopies

```agda
module _
  {l1 l2 : Level} {A : UU l1} {B : A → UU l2} {f g h u v i : (x : A) → B x}
  where

  double-whisker-coherence-square-homotopies :
    (p : f ~ g)
    (top : g ~ u) (left : g ~ h) (right : u ~ v) (bottom : h ~ v)
    (s : v ~ i) →
    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies
      ( p ∙h top)
      ( p ∙h left)
      ( right ∙h s)
      ( bottom ∙h s)
  double-whisker-coherence-square-homotopies p top left right bottom q H =
    left-whisker-concat-coherence-square-homotopies p top left
      ( right ∙h q)
      ( bottom ∙h q)
    ( right-whisker-concat-coherence-square-homotopies
      ( top)
      ( left)
      ( right)
      ( bottom)
      ( H)
      ( q))
```

#### Left splicing coherences of commuting squares of homotopies

> For any inverse pair of homotopies `H : g ~ u` and `K : u ~ g` equipped with
> `α : inv-htpy H ~ K` we obtain an equivalence
> 
> ```text
>            top                                    top
>        f -------> g                         f -----------> g
>        |          |                         |              |
>   left |          | right    ≃    left ∙h H |              | right
>        ∨          ∨                         ∨              ∨
>        h -------> i                         u -----------> i
>           bottom                               K ∙h bottom
> ```
> 
> of coherences of commuting squares of homotopies.

Reviewer: ```suggestion
For any inverse pair of homotopies `H : g ~ u` and `K : u ~ g` equipped with
`α : inv-htpy H ~ K` we obtain maps back and forth

```text
           top                                    top
       f -------> g                         f -----------> g
       |          |                         |              |
  left |          | right    ↔    left ∙h H |              | right
       ∨          ∨                         ∨              ∨
       h -------> i                         u -----------> i
          bottom                               K ∙h bottom
```

of coherences of commuting squares of homotopies. We show in ... that these maps are equivalences.
```

```agda
module _
  {l1 l2 : Level} {A : UU l1} {B : A → UU l2} {f g h i : (x : A) → B x}
  (top : f ~ g) (left : f ~ h) (right : g ~ i) (bottom : h ~ i)
  where

  left-splice-coherence-square-homotopies :
    {u : (x : A) → B x} (H : h ~ u) (K : u ~ h) (α : inv-htpy H ~ K) →
    coherence-square-homotopies top left right bottom →
    coherence-square-homotopies top (left ∙h H) right (K ∙h bottom)
  left-splice-coherence-square-homotopies H K α coh x =
    left-splice-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( H x)
      ( K x)
      ( α x)
      ( coh x)

  left-unsplice-coherence-square-homotopies :
    {u : (x : A) → B x} (H : h ~ u) (K : u ~ h) (α : inv-htpy H ~ K) →
    coherence-square-homotopies top (left ∙h H) right (K ∙h bottom) →
    coherence-square-homotopies top left right bottom
  left-unsplice-coherence-square-homotopies H K α coh x =
    left-unsplice-coherence-square-identifications
      ( top x)
      ( left x)
      ( right x)
      ( bottom x)
      ( H x)
      ( K x)
      ( α x)
      ( coh x)
```

#### Right splicing coherences of commuting squares of homotopies

> For any inverse pair of homotopies `H : g ~ u` and `K : u ~ g` equipped with
> `α : inv-htpy H ~ K` we obtain an equivalence
> 
> ```text
>            top                             top ∙h H
>        f -------> g                     f --------> u
>        |          |                     |           |
>   left |          | right    ≃     left |           | K ∙h right
>        ∨          ∨                     ∨           ∨
>        h -------> i                     h --------> i
>           bottom                           bottom
> ```
> 
> of coherences of commuting squares of homotopies.

Reviewer: ````suggestion
For any inverse pair of homotopies `H : g ~ u` and `K : u ~ g` equipped with
`α : inv-htpy H ~ K` we obtain maps back and forth

```text
           top                             top ∙h H
       f -------> g                     f --------> u
       |          |                     |           |
  left |          | right    ↔     left |           | K ∙h right
       ∨          ∨                     ∨           ∨
       h -------> i                     h --------> i
          bottom                           bottom
```

of coherences of commuting squares of homotopies. We show in ... that these maps are equivalences.
````
